### PR TITLE
[release 148t] Fix comparison for char such as padding is taken into account

### PR DESCRIPTION
**Pull request by maciejgrzybek**
Based on patch reviewed here: https://github.com/Teradata/presto/pull/363

### CODE REVIEW

**Comment by maciejgrzybek** [comment]
Disputes about `char.no-pad` flag resolved in offline discussion with @arhimondr.

File: presto-main/src/main/java/com/facebook/presto/metadata/FunctionRegistry.java [hunk 1]
```diff
@@ -531,24 +524,13 @@ public WindowFunctionSupplier load(SpecializedFunctionKey key)
...
-        else {
-            builder.function(CHAR_EQUAL_NO_PAD)
-                    .function(CHAR_NOT_EQUAL_NO_PAD)
-                    .function(CHAR_LESS_THAN_NO_PAD)
-                    .function(CHAR_LESS_THAN_OR_EQUAL_NO_PAD)
-                    .function(CHAR_GREATER_THAN_NO_PAD)
-                    .function(CHAR_GREATER_THAN_OR_EQUAL_NO_PAD)
-                    .function(CHAR_BETWEEN_NO_PAD);
-        }
+        builder.function(CHAR_EQUAL_PAD_SPACES)
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/metadata/FunctionRegistry.java:527]: Shouldn't NO PAD version be the used one? I know that char coercions in 148 don't allow for NO PAD version to be implemented correctly in joins. However, in 148 NO PAD was used by default, so not we are changing behavior for our client. Is that OK for client? If OK, then I am fine with this commit.

File: presto-main/src/test/java/com/facebook/presto/type/TestCharOperators.java [hunk 2]
```diff
@@ -30,6 +30,9 @@ public void testEqual()
         assertFunction("cast('foo' as char(5)) = cast('foo' as char(3))", BOOLEAN, true);
         assertFunction("cast('foo' as char(3)) = cast('bar' as char(3))", BOOLEAN, false);
         assertFunction("cast('bar' as char(3)) = cast('foo' as char(3))", BOOLEAN, false);
+
+        assertFunction("cast('a' as char(2)) = cast('a ' as char(2))", BOOLEAN, true);
```
Comment (sopel39) [inline presto-main/src/test/java/com/facebook/presto/type/TestCharOperators.java]: add test case for 'a a' comapred with 'a ' and also \0 cases?

**Comment by sopel39** [comment]
`Remove "char.no-pad" flag` changes default CHAR comparison behavior in 148t. If that is OK for client, then LGTM. FYI: @KBP-TDC

**Comment by fiedukow** [comment]
Just to clarify, changing this default behavior will cause transition between CHAR comparison behavior twice (assuming there will be update to 0.152-t or newer version).
- 0.148-t: `cast('a' as char(2)) = cast('a' as char(1)) # false`
- 0.148-t with this path: `cast('a' as char(2)) = cast('a' as char(1)) # true`
- 0.152-t: `cast('a' as char(2)) = cast('a' as char(1)) # false`

I just checked all three scenarios in all three versions just to be sure.

So this is the side effect of this patch in current form. This may be tradeoff we want to accept but it's nice to be fully aware of that.

**Comment by maciejgrzybek** [comment]
As per offline discussion with @KBP-TDC, we've decided to keep this patch backward compatible with 148t (I'll prepare a fixup).
There will be a separate ticket to document current behavior and to decide on proper semantics if need to change anything.

**Comment by maciejgrzybek** [comment]
FYI:
PSQL:

``` sql
# select cast('a' as char(2)) = cast('a' as char(1));
 ?column? 
----------
 t
(1 row)
```

``` sql
# select cast('a ' as char(2)) = cast('a' as char(1));
 ?column? 
----------
 t
(1 row)
```

**Comment by maciejgrzybek** [comment]
Please move to https://github.com/Teradata/presto/pull/370.

**Comment by fiedukow** [comment]
The PSQL argument is totally right, we were opting for that as well (this is only valid ANSI SQL behavior btw).
The problem is that the current solution is the compromise with FB point of view, so even though it's not aligned with SQL standard it may be in prestodb/presto for long time and we probably don't want to differentiate that in our distribution. I think this leads us to at least being consequent with our own behavior.

**Pull request closed without merging**